### PR TITLE
feat: subgraph error propagation improvements

### PR DESCRIPTION
**Pull request by StarpTech**
## Motivation and Context

This PR comes with the following improvements:

- Add additional options to configure error propagation:
  - `default_extension_code`: (Default: `DOWNSTREAM_SERVICE_ERROR`)The default extension code. The default extension code is used to specify the default code for the Subgraph errors when the code is not present.
  - `allowed_extension_fields`:  (Default: `"code"`) The allowed extension fields. The allowed extension fields are used to specify which fields of the Subgraph errors are allowed to be propagated to the client. In development mode, we additionally allow `stacktrace` to be part of the extensions.
  - `attach_service_name`: (Default: `true`) Attach the service name to each Subgraph error. If the value is true (default), the service name will be attached to the Subgraph errors.
- Set proper `application/json` content type in errors that aren't written by the engine.
- Don't set `data:null` for errors that are happening before query execution. This compliant with the [spec](https://spec.graphql.org/draft/#sec-Data).
- Use subgraph name in error messages.
- Deprecate `subgraph_error_propagation.enabled` and make it true by default.

Requires: https://github.com/wundergraph/graphql-go-tools/pull/883

The following configuration can be used to get the most common propagation behaviour:
```yaml
subgraph_error_propagation:
  mode: pass-through
  omit_locations: false
  default_extension_code: DOWNSTREAM_SERVICE_ERROR
  attach_service_name: true
  allowed_extension_fields:
    - "code"
```

## Example

```json
{
    "errors": [
        {
            "message": "error resolving RootFieldThrowsError for Employee 12",
            "path": [
                "employees",
                9,
                "rootFieldThrowsError"
            ],
            "extensions": {
                "code": "ERROR_CODE",
                "serviceName": "employees",
                "statusCode": 200
            }
        }
    ],
    "data": null
}
```

## TODO

- [ ] Tests or benchmark included
- [ ] Documentation is changed or added on [https://app.gitbook.com/](https://app.gitbook.com/)
- [ ] PR title must follow [conventional-commit-standard](https://github.com/wundergraph/wundergraph/blob/main/CONTRIBUTING.md#conventional-commit-standard)

### CODE REVIEW

File: router/core/errors.go [hunk 1]
```diff
@@ -122,10 +122,14 @@ func propagateSubgraphErrors(ctx *resolve.Context, logger *zap.Logger) {
 // It accepts a graphqlerrors.RequestErrors object and writes it to the response based on the GraphQL spec.
 func writeRequestErrors(r *http.Request, w http.ResponseWriter, statusCode int, requestErrors graphqlerrors.RequestErrors, requestLogger *zap.Logger) {
 	if requestErrors != nil {
-		if statusCode != 0 {
-			w.WriteHeader(statusCode)
-		}
+
 		if r.URL.Query().Has("wg_sse") {
```
Comment (jensneuse) [inline router/core/errors.go:126]: did you not want to fix this? =)

Comment (StarpTech) [reply router/core/errors.go:126]: Yes, but in a separate PR. This is already too big.

File: demo/pkg/subgraphs/employees/subgraph/schema.resolvers.go [hunk 2]
```diff
@@ -10,13 +10,23 @@ import (
...
 )
 
 // RootFieldThrowsError is the resolver for the rootFieldThrowsError field.
 func (r *employeeResolver) RootFieldThrowsError(ctx context.Context, obj *model.Employee) (*string, error) {
-	return nil, fmt.Errorf("error resolving RootFieldThrowsError for Employee %d", obj.ID)
+	graphql.AddError(ctx, &gqlerror.Error{
+		Path:    graphql.GetPath(ctx),
+		Message: fmt.Sprintf("error resolving RootFieldThrowsError for Employee %d", obj.ID),
+		Extensions: map[string]interface{}{
+			"code": "ERROR_CODE",
```
Comment (Aenimus) [inline demo/pkg/subgraphs/employees/subgraph/schema.resolvers.go:24]: Is this intended or debug code?

File: router-tests/feature_flags_test.go [hunk 3]
```diff
@@ -20,7 +20,7 @@ func TestFeatureFlags(t *testing.T) {
 				Query: `{ employees { id productCount } }`,
 			})
 			require.Empty(t, res.Response.Header.Get("X-Feature-Flag"))
-			require.JSONEq(t, `{"errors":[{"message":"field: productCount not defined on type: Employee","path":["query","employees","productCount"]}],"data":null}`, res.Body)
```
Comment (Aenimus) [inline router-tests/feature_flags_test.go:23]: So if there is a propagated error, the data shouldn't be null—the property should not exist at all?

Comment (StarpTech) [reply router-tests/feature_flags_test.go:23]: All errors before query execution e.g parsing, validation should not contain data property. https://spec.graphql.org/draft/#sec-Data

**Review by Aenimus — APPROVED** [review]
LGTM!

**Pull request merged**